### PR TITLE
[RISCV][GISEL] Legalize and regbankselect vector typed G_IMPLICIT_DEF

### PR DESCRIPTION
**Pull request by michaelmaitland**
Let instruction selection do what it wants with this opcode for scalable vector types that are already legal.

### CODE REVIEW

**Comment by tschuett** [comment]
Did you try legalization of illegal types, e.g.,  LLT::scalable_vector(30, s1)?

**Comment by tschuett** [comment]
Disallow illegal types! But the legalizer should transform illegal types into legal types. For scalable vectors there is not yet good test coverage.

**Comment by michaelmaitland** [comment]
> Disallow illegal types! But the legalizer should transform illegal types into legal types. For scalable vectors there is not yet good test coverage.

I have added a TODO in the legalizer to make illegal types of G_IMPLICIT_DEF into legal types, and I have updated this PR to specify that this patch only legalizes vector types that are considered legal.

**Comment by michaelmaitland** [comment]
ping

File: llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp [hunk 1]
```diff
@@ -351,6 +351,11 @@ RISCVRegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
     // $f10_d = COPY %1(s32)
     if (anyUseOnlyUseFP(Dst, MRI, TRI))
       Mapping = getFPValueMapping(MRI.getType(Dst).getSizeInBits());
+
+    LLT DstTy = MRI.getType(Dst);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:355]: Move the declaration of `LLT DstTy` earlier and use it on line 353 too.

**Comment by michaelmaitland** [comment]
ping, all my open GISEL PRs depend on this one :)

**Review by topperc — APPROVED** [review]
LGTM

**Comment by michaelmaitland** [comment]
Committed in 33cf27b0a3f307f4c57e868fedb72f4d65ba72d9 and ea798a79000f5df53b469ecb491d7d21479d6f24

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 2]
```diff
@@ -40,6 +40,30 @@ static LegalityPredicate typeIsScalarFPArith(unsigned TypeIdx,
...
+  return all(typeInSet(TypeIdx, IntOrFPVecTys), P);
+}
+
+static LegalityPredicate
+typeIsLegalBoolVec(unsigned TypeIdx, std::initializer_list<LLT> BoolVecTys,
+                   const RISCVSubtarget &ST) {
+  LegalityPredicate HasV = [=, &ST](const LegalityQuery &Query) {
+    return ST.hasVInstructions();
+  };
+  return all(typeInSet(TypeIdx, BoolVecTys), HasV);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:64]: I missed it in the review, but this part of regular vectors applies to bools too

```
           (Query.Types[TypeIdx].getElementCount().getKnownMinValue() != 1 ||    
            ST.getELen() == 64); 
```

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:64]: I will add it

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:64]: 3305f5a980363d7e224094bc42bcbebcf162618e

**Pull request closed without merging**